src/profile/hooks/useProfileDataEditor.ts

```diff
@@ -1,34 +1,109 @@
-import { useContext, useMemo } from 'react';
+import { useEffect, useMemo, useState } from 'react';
+import to from 'await-to-js';
 
-import { ProfileContext } from '../context/ProfileContext';
 import {
   EditData,
   EditDataType,
   createEditorForDataType,
+  SaveType,
+  EditDataValue,
+  FormValues,
 } from '../helpers/editData';
+import { UpdateResult, useProfileMutations } from './useProfileMutations';
 
 export type ProfileDataEditorReturnType = {
   editDataList: EditData[];
+  save: (item: EditData, newValue: EditDataValue) => Promise<UpdateResult>;
+  reset: (item: EditData) => void;
 };
 
+export type Action = SaveType | 'edit' | 'cancel' | 'save' | 'add';
+
+export type ActionListenerReturnType = Promise<void | UpdateResult>;
+export type ActionListener = (
+  action: Action,
+  data: EditData,
+  value?: Partial<EditDataValue>
+) => ActionListenerReturnType;
+
+export function saveTypeToAction(saveType?: SaveType): Action | undefined {
+  if (!saveType) {
+    return undefined;
+  }
+  if (saveType === 'value') {
+    return 'save';
+  }
+  if (saveType === 'remove' || saveType === 'set-primary') {
+    return saveType;
+  }
+  return undefined;
+}
+
 export function useProfileDataEditor({
   dataType,
 }: {
   dataType: EditDataType;
 }): ProfileDataEditorReturnType {
-  const { data: profileData } = useContext(ProfileContext);
+  const { profileData, update: updateProfile } = useProfileMutations({
+    dataType,
+  });
   if (!profileData) {
     throw new Error(
       'No profile data for useProfileDataEditor. Hook used before data has been fetched.'
     );
   }
-  const { getEditData } = useMemo(
+  const {
+    getEditData,
+    updateItemAndCreateSaveData,
+    updateData,
+    updateAfterSavingError,
+    resetItem,
+  } = useMemo(
     () => createEditorForDataType(profileData, dataType),
     // eslint-disable-next-line react-hooks/exhaustive-deps
     []
   );
 
+  const [, setUpdateTime] = useState<number>(Date.now());
+  const triggerUpdate = () => {
+    setUpdateTime(Date.now());
```

Comment:
Not sure what is the purpose of this triggerUpdate? Is it to just manually trigger UI update by applying useState? It kind of feels like a hacky thing to do. I wonder if there's any alternatives 🤔

Reply:
Yes, when data is updated in some way that does not make hook re-render (like pressing cancel) the re-render must be triggered manually. Same kind of easy re-render is used for example in MobX:

https://github.com/mobxjs/mobx-react-lite/blob/master/src/utils/utils.ts